src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(chip::Messaging::ExchangeContext * ec,
+                                                       const chip::PayloadHeader & payloadHeader,
+                                                       chip::System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    ChipLogDetail(BDX, " %s: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId, __FUNCTION__,
+                  payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+
+    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
+    ec->WillSendMessage();
+
+    // Get the next output event and send it to the delegate.
+    TransferSession::OutputEvent outEvent;
+    do
+    {
+        mTransfer.GetNextAction(outEvent);
+
+        if (outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout)
+        {
+            OnResponseTimeout(ec);
+        }
+        else
+        {
+            HandleAsyncTransferSessionOutput(outEvent);
+        }
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
```

Comment:
I assume you copied this from BDXDownloader.cpp, but why are we delivering `TransferSession::OutputEventType::kNone` events to our client?  I would think this logic should be more like:

1. Get event.
2. If it's kNone, break out of the loop.
3. Deliver the event to the client.

Reply:
i can do that.

Reply:
done.